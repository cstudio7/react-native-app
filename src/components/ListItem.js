import React from 'react';
import { Text, View, TouchableOpacity, StyleSheet } from 'react-native';
import { Icon } from 'expo';

class ListItem extends React.Component {
  render() {
    const { name } = this.props;

    return (
      <View style={styles.listItem}>
        <Text style={styles.listText}>{name.name}</Text>
        <TouchableOpacity
          onPress={async () => {
            const payload = {
              gender: this.props.gender,
              name: {
                ...name,
                isFavorite: !name.isFavorite
              }
            };
            if (name.isFavorite) {
              this.props.unfavoriteName(payload);
            } else {
              this.props.favoriteName(payload);
            }
          }}>
          <Icon.Ionicons
            name={`md-heart${name.isFavorite ? '' : '-outline'}`}
            style={styles.icon}
          />
        </TouchableOpacity>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  listItem: {
    borderTopWidth: 1,
    borderColor: 'rgba(0, 0, 0, 0.15)',
    paddingTop: 22,
    paddingBottom: 22,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  listText: {
    fontSize: 19
  },
  icon: {
    fontSize: 20,
    color: '#7A62A4',
    marginRight: 16,
    marginTop: 2
  }
});

export default ListItem;
