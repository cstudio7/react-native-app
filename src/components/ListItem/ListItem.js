import React from 'react';
import {
  Text,
  View,
  TouchableOpacity,
  TouchableWithoutFeedback,
  StyleSheet
} from 'react-native';
import { Icon } from 'expo';
import Spacing from '../../constants/Spacing';

class ListItem extends React.PureComponent {
  render() {
    const { name, favorite, openNameScreen, gender } = this.props;

    return (
      <TouchableWithoutFeedback onPress={openNameScreen}>
        <View style={styles.listItem}>
          <Text style={styles.listText}>{name.name}</Text>
          <TouchableOpacity
            style={styles.touchableArea}
            onPress={() => {
              const payload = {
                gender,
                name: {
                  ...name,
                  isFavorite: !name.isFavorite
                }
              };
              favorite(payload);
            }}>
            <Icon.Ionicons
              name={`md-heart${name.isFavorite ? '' : '-outline'}`}
              style={styles.icon}
            />
          </TouchableOpacity>
        </View>
      </TouchableWithoutFeedback>
    );
  }
}

const styles = StyleSheet.create({
  listItem: {
    borderTopWidth: 1,
    borderColor: 'rgba(0, 0, 0, 0.15)',
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  listText: {
    fontSize: 19,
    marginTop: Spacing.margin3,
    marginBottom: Spacing.margin3
  },
  icon: {
    fontSize: 22,
    color: '#7A62A4',
    marginRight: Spacing.base
  },
  touchableArea: {
    paddingLeft: Spacing.padding3,
    paddingTop: Spacing.padding3,
    paddingBottom: Spacing.padding3
  }
});

export default ListItem;
