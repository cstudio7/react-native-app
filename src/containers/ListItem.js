import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import ListItem from '../components/ListItem/ListItem';
import { favorite } from '../redux/namesReducer/actions';

const mapStateToProps = () => ({});

const mapDispatchToProps = (dispatch, props) => {
  return bindActionCreators(
    {
      onPress: favorite.bind(null, props.screen)
    },
    dispatch
  );
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListItem);
