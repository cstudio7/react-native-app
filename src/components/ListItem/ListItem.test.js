import React from 'react';
import { TouchableOpacity } from 'react-native';
import renderer from 'react-test-renderer';
import ListItem from './ListItem';

const name = { id: 1, name: 'artur' };

it('renders a name correctly', () => {
  const tree = renderer.create(<ListItem name={name} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a favorite name correctly', () => {
  const favoriteName = {
    ...name,
    isFavorite: true
  };
  const tree = renderer.create(<ListItem name={favoriteName} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('triggers onPress when the user taps', () => {
  const onPressEvent = jest.fn();
  const page = 'favorites';
  const gender = 'male';
  const expectedPayload = {
    gender,
    name: {
      ...name,
      isFavorite: true // since we've triggered onPress
    }
  };

  const instance = renderer.create(
    <ListItem
      name={name}
      gender={gender}
      onPress={onPressEvent}
      page={page}
    />
  ).root;

  instance.findByType(TouchableOpacity).props.onPress();

  expect(onPressEvent).toHaveBeenCalledWith(expectedPayload, page);
});
