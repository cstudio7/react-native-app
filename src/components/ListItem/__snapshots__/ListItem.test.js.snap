// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a favorite name correctly 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "rgba(0, 0, 0, 0.15)",
      "borderTopWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 19,
        "marginBottom": 24,
        "marginTop": 24,
      }
    }
  >
    artur
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "paddingBottom": 24,
        "paddingLeft": 24,
        "paddingTop": 24,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
  </View>
</View>
`;

exports[`renders a name correctly 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "rgba(0, 0, 0, 0.15)",
      "borderTopWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 19,
        "marginBottom": 24,
        "marginTop": 24,
      }
    }
  >
    artur
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "paddingBottom": 24,
        "paddingLeft": 24,
        "paddingTop": 24,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
  </View>
</View>
`;
